server/list.js

```diff
@@ -102,48 +102,57 @@ function getOptions(id) {
   }
 }
 
-async function fetchAllFiles({nextPageToken: pageToken, listSoFar = [], parentIds = [driveId], driveType = 'team', drive} = {}) {
-  const options = getOptions(parentIds)
+async function fetchAllFiles({nextPageToken: pageToken, nextListSoFar: listSoFar = [], nextParentId: parentId = driveId, nextParentName: parentName = 'root', driveType = 'team', drive} = {}) {
+  const options = getOptions(parentId)
 
   if (pageToken) {
     options.pageToken = pageToken
   }
 
-  log.debug(`searching for files > ${listSoFar.length}`)
+  log.debug(`searching for files > ${listSoFar.length} in "${parentName}"`)
 
   // Gets files in single folder (shared) or files listed in single page of response (team)
   const {data} = await drive.files.list(options)
 
   const {files, nextPageToken} = data
-  const combined = listSoFar.concat(files)
+  const sublist = []
 
-  // If there is more data the API has not returned for the query, the request needs to continue
-  if (nextPageToken) {
-    return fetchAllFiles({
-      nextPageToken,
-      listSoFar: combined,
-      drive
-    })
+  if (driveType === 'folder') {
+    // Continue searching if shared folder, since API only returns contents of the immediate parent folder
+    // Find folders that have not yet been searched
+    const folders = files.filter((item) =>
+      item.mimeType === 'application/vnd.google-apps.folder' && parentId === item.parents[0])
+
+    if (folders.length > 0) {
+      // Avoid Promise.all to avoid hitting google API rate limits
+      await folders.reduce((acc, folder) => {
+        return acc.then((files) => fetchAllFiles({
+          drive,
+          driveType: 'folder',
+          sublist,
+          nextPageToken,
+          nextParentId: folder.id,
+          nextParentName: folder.name
+        }).then((newFiles) => files.concat(newFiles)))
+      }, Promise.resolve([]))
+    }
   }
 
-  // If there are no more pages and this is not a shared folder, return completed list
-  if (driveType !== 'folder') return combined
-
-  // Continue searching if shared folder, since API only returns contents of the immediate parent folder
-  // Find folders that have not yet been searched
-  const folders = combined.filter((item) =>
-    item.mimeType === 'application/vnd.google-apps.folder' && parentIds.includes(item.parents[0]))
+  const nextListSoFar = listSoFar.concat(files).concat(sublist)
 
-  if (folders.length > 0) {
+  // If there is more data the API has not returned for the query, the request needs to continue
+  if (nextPageToken) {
     return fetchAllFiles({
-      listSoFar: combined,
       drive,
-      parentIds: folders.map((folder) => folder.id),
-      driveType
+      driveType,
+      nextListSoFar,
+      nextPageToken,
+      nextParentId: parentId,
+      nextParentName: parentName
```

Comment:
this was the underlying issue, when paginating through a large folder, parentIds wasn't being propogated so we would jump back up to the root and the request would fail